libraries/ui/src/main/java/androidx/media3/ui/SubtitlePainter.java

```diff
@@ -470,4 +476,117 @@ private static boolean areCharSequencesEqual(
     // equals methods, so we perform one explicitly here.
     return first == second || (first != null && first.equals(second));
   }
+
+  /**
+   * Checks whether the given {@link CharSequence} contains any characters
+   * with right-to-left (RTL) directionality.
+   * <p>
+   * This method inspects each character's Unicode directionality and returns {@code true}
+   * if at least one character is classified as RTL. This includes characters with the following
+   * directionalities:
+   * <ul>
+   *     <li>{@link Character#DIRECTIONALITY_RIGHT_TO_LEFT}</li>
+   *     <li>{@link Character#DIRECTIONALITY_RIGHT_TO_LEFT_ARABIC}</li>
+   *     <li>{@link Character#DIRECTIONALITY_RIGHT_TO_LEFT_EMBEDDING}</li>
+   *     <li>{@link Character#DIRECTIONALITY_RIGHT_TO_LEFT_OVERRIDE}</li>
+   * </ul>
+   *
+   * @param input the input {@link CharSequence} to analyze
+   * @return {@code true} if the input contains at least one RTL character; {@code false} otherwise
+   */
+  private static boolean containsRTL(@Nullable CharSequence input) {
+    if (input == null) {
+      return false;
+    }
+    int length = input.length();
+    for (int offset = 0; offset < length; ) {
+      int codePoint = Character.codePointAt(input, offset);
+      byte dir = Character.getDirectionality(codePoint);
+      if (dir == Character.DIRECTIONALITY_RIGHT_TO_LEFT ||
+          dir == Character.DIRECTIONALITY_RIGHT_TO_LEFT_ARABIC ||
+          dir == Character.DIRECTIONALITY_RIGHT_TO_LEFT_EMBEDDING ||
+          dir == Character.DIRECTIONALITY_RIGHT_TO_LEFT_OVERRIDE) {
+        return true;
+      }
+      offset += Character.charCount(codePoint);
+    }
+    return false;
+  }
+
+  /**
+   * Applies bidirectional (bidi) Unicode wrapping to each line of the given {@link CharSequence}.
+   * <p>
+   * This method ensures that text containing both left-to-right (LTR) and right-to-left (RTL)
+   * scripts is displayed correctly by wrapping each line using {@link BidiFormatter#unicodeWrap}.
+   * It forces LTR context for wrapping and preserves spans and line breaks.
+   *
+   * @param input the input text as a {@link CharSequence}, possibly containing mixed-direction text
+   * @return a {@link CharSequence} with each line wrapped for proper bidi rendering
+   */
+  public static CharSequence wrapBidiText(CharSequence input) {
+    BidiFormatter bidiFormatter = BidiFormatter.getInstance();
+
+    // Preserve span in the input text.
+    Spanned spannedInput = (Spanned)input;
```

Comment:
This cast should be conditional on an `instanceof` check (i.e. you should also support cues where the text is just a `String` and not a `Spanned`)